util/stmtsummary/statement_summary.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stmtsummary
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pingcap/parser/mysql"
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/kvcache"
+)
+
+type stmtSummaryCacheKey struct {
+	// Same statements may appear in different schema, but they refer to different tables.
+	schemaName string
+	digest     string
+	// TODO: add plan digest
+	// `hash` is the hash value of this object
+	hash []byte
+}
+
+// Hash implements SimpleLRUCache.Key
+func (key *stmtSummaryCacheKey) Hash() []byte {
+	if len(key.hash) == 0 {
+		key.hash = make([]byte, 0, len(key.schemaName)+len(key.digest))
+		key.hash = append(key.hash, hack.Slice(key.digest)...)
+		key.hash = append(key.hash, hack.Slice(strings.ToLower(key.schemaName))...)
+	}
+	return key.hash
+}
+
+// A LRU cache that stores statement summary.
+type stmtSummaryByDigest struct {
+	sync.RWMutex
+	summaryMap *kvcache.SimpleLRUCache
+}
+
+// StmtSummary is the global object for statement summary.
+var StmtSummary = NewStmtSummaryByDigest()
+
+// Summary of each type of statements.
+type stmtSummary struct {
+	schemaName      string
+	digest          string
+	normalizedSQL   string
+	sampleSQL       string
+	execCount       uint64
+	sumLatency      uint64
+	maxLatency      uint64
+	minLatency      uint64
+	sumAffectedRows uint64
+	// Number of rows sent to client.
+	sumSentRows uint64
+	// The first time this type of SQL executes.
+	firstSeen time.Time
+	// The last time this type of SQL executes.
+	lastSeen time.Time
+}
+
+// StmtExecInfo records execution status of each statement.
+type StmtExecInfo struct {
+	SchemaName    string
+	OriginalSQL   string
+	NormalizedSQL string
+	Digest        string
+	TotalLatency  uint64
+	AffectedRows  uint64
+	// Number of rows sent to client.
+	SentRows  uint64
+	StartTime time.Time
+}
+
+// NewStmtSummaryByDigest creates a stmtSummaryByDigest.
+func NewStmtSummaryByDigest() *stmtSummaryByDigest {
+	maxStmtCount := config.GetGlobalConfig().StmtSummary.MaxStmtCount
+	return &stmtSummaryByDigest{
+		summaryMap: kvcache.NewSimpleLRUCache(maxStmtCount, 0, 0),
```

Comment:
Is it possible that we still need memory quota, say, if the maxStmtCount is set to be very very large?

Reply:
It complicates the configuration of statement summary, which makes users more confusing. If the user sets max-stmt-count to a very large value, he can also set the quota to a very large value.